### PR TITLE
Intellisense and Union/Coercion improvements for Derived Record Types

### PR DESCRIPTION
**Pull request by lesaltzm**
Attempting to integrate ControlTypes with Lazy Records showed a set of gaps in the existing Derived RecordType support. This closes some of those holes.

### CODE REVIEW

**Comment by BruceHaley** [comment]
:heavy_check_mark: No Binary Compatibility issues for **Microsoft.PowerFx.Connectors.dll**
:heavy_check_mark: No Binary Compatibility issues for **Microsoft.PowerFx.Core.dll**
:heavy_check_mark: No Binary Compatibility issues for **Microsoft.PowerFx.Interpreter.dll**
:heavy_check_mark: No Binary Compatibility issues for **Microsoft.PowerFx.LanguageServerProtocol.dll**
:heavy_check_mark: No Binary Compatibility issues for **Microsoft.PowerFx.Transport.Attributes.dll**

**Comment by BruceHaley** [comment]
:heavy_check_mark: No Binary Compatibility issues for **Microsoft.PowerFx.Connectors.dll**
:heavy_check_mark: No Binary Compatibility issues for **Microsoft.PowerFx.Core.dll**
:heavy_check_mark: No Binary Compatibility issues for **Microsoft.PowerFx.Interpreter.dll**
:heavy_check_mark: No Binary Compatibility issues for **Microsoft.PowerFx.LanguageServerProtocol.dll**
:heavy_check_mark: No Binary Compatibility issues for **Microsoft.PowerFx.Transport.Attributes.dll**

File: src/libraries/Microsoft.PowerFx.Core/Types/DType.cs [hunk 1]
```diff
@@ -966,41 +966,6 @@ public string GetKindString()
...
-        // control-specific information, such as property names.
-        // Reasons this is bad:
-        //  1. It is recursive, and will go as deep into a DType as needed to convert all nested
-        //     controls to their corresponding record representations.
-        //  2. It converts "v" (control) types to records, by picking certain property names as
-        //     fields for those records. This operation is LOSSY. For example, only the locale-specific
-        //     names are captured, not the invariant names (or vice versa).
-        //  3. There is no way to recover the originating control or control template from the
-        //     resulting type.
-        public DType ControlsToRecordsRecursive()
```
Comment (MikeStall) [inline src/libraries/Microsoft.PowerFx.Core/Types/DType.cs:981]: >ControlsToRecordsRecursive [](http://example.com/codeflow?start=21&length=26)

removing this is great!

**Comment by BruceHaley** [comment]
:heavy_check_mark: No Binary Compatibility issues for **Microsoft.PowerFx.Connectors.dll**
:heavy_check_mark: No Binary Compatibility issues for **Microsoft.PowerFx.Core.dll**
:heavy_check_mark: No Binary Compatibility issues for **Microsoft.PowerFx.Interpreter.dll**
:heavy_check_mark: No Binary Compatibility issues for **Microsoft.PowerFx.LanguageServerProtocol.dll**
:heavy_check_mark: No Binary Compatibility issues for **Microsoft.PowerFx.Transport.Attributes.dll**

**Review by MikeStall — APPROVED** [review]
:shipit:

File: src/libraries/Microsoft.PowerFx.Core/Types/DType.cs [hunk 2]
```diff
@@ -2589,6 +2554,11 @@ public static DType Union(ref bool fError, DType type1, DType type2, bool useLeg
             // For Lazy Types, union operations must expand the current depth
             if (type1.IsLazyType)
             {
+                if (type1 == type2)
```
Comment (Shpakh-MSFT) [inline src/libraries/Microsoft.PowerFx.Core/Types/DType.cs:2557]: >type1 == type2 [](http://example.com/codeflow?start=20&length=14)

What's the definition of equals for lazytype here? Is it purely referential?

Comment (lesaltzm) [reply src/libraries/Microsoft.PowerFx.Core/Types/DType.cs:2557]: Implemented through the LazyTypeProvider

**Review by Shpakh-MSFT — APPROVED** [review]
:shipit:

**Comment by BruceHaley** [comment]
:heavy_check_mark: No Binary Compatibility issues for **Microsoft.PowerFx.Connectors.dll**
:heavy_check_mark: No Binary Compatibility issues for **Microsoft.PowerFx.Core.dll**
:heavy_check_mark: No Binary Compatibility issues for **Microsoft.PowerFx.Interpreter.dll**
:heavy_check_mark: No Binary Compatibility issues for **Microsoft.PowerFx.LanguageServerProtocol.dll**
:heavy_check_mark: No Binary Compatibility issues for **Microsoft.PowerFx.Transport.Attributes.dll**

**Review by bryancroteau-MSFT — APPROVED** [review]

**Pull request merged**